streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

```diff
@@ -102,14 +127,98 @@ public void process(final K key, final V1 value) {
+                    //     The record is not processed yet, and is added to the outer-join store
+                    //  3. A joined record with time T11 is seen in the left-topic (maxLeftStreamTime: 11)
+                    //     It is time to look at the expired records. T10 and T2 should be emitted, but
+                    //     because T2 was late, then it is not fetched by the window store, so it is not processed
+                    //
+                    // See KStreamKStreamLeftJoinTest.testLowerWindowBound() tests
+                    //
+                    // the condition below allows us to process the late record without the need
+                    // to hold it in the temporary outer store
+                    if (timeTo < maxStreamTime) {
+                        context().forward(key, joiner.apply(key, value, null));
+                    } else {
+                        outerJoinWindowStore.ifPresent(store -> store.put(
+                            KeyAndJoinSide.make(thisJoin, key),
+                            makeValueOrOtherValue(thisJoin, value),
+                            inputRecordTimestamp));
+                    }
+                }
+
+                outerJoinWindowStore.ifPresent(store -> {
+                    // only emit left/outer non-joined if the stream time has advanced (inputRecordTime = maxStreamTime)
+                    // if the current record is late, then there is no need to check for expired records
+                    if (inputRecordTimestamp == maxStreamTime) {
+                        maybeEmitOuterExpiryRecords(store, maxStreamTime);
+                    }
+                });
+            }
+        }
+
+        private ValueOrOtherValue makeValueOrOtherValue(final boolean thisJoin, final V1 value) {
+            return thisJoin
+                ? ValueOrOtherValue.makeValue(value)
+                : ValueOrOtherValue.makeOtherValue(value);
+        }
+
+        @SuppressWarnings("unchecked")
+        private void maybeEmitOuterExpiryRecords(final WindowStore<KeyAndJoinSide<K>, ValueOrOtherValue> store, final long maxStreamTime) {
+            try (final KeyValueIterator<Windowed<KeyAndJoinSide<K>>, ValueOrOtherValue> it = store.all()) {
+                while (it.hasNext()) {
+                    final KeyValue<Windowed<KeyAndJoinSide<K>>, ValueOrOtherValue> e = it.next();
+
+                    // Skip next records if the oldest record has not expired yet
+                    if (e.key.window().end() + joinGraceMs >= maxStreamTime) {
+                        break;
+                    }
+
+                    final K key = e.key.key().getKey();
+
+                    // Emit the record by joining with a null value. But the order varies depending whether
+                    // this join is using a reverse joiner or not. Also whether the returned record from the
+                    // outer window store is a V1 or V2 value.
+                    if (thisJoin) {
```

Comment:
Here if we refactor to `left / right` then this logic can be simplified as well since we would only care whether the deserialized key/value are left or right.

Reply:
Done